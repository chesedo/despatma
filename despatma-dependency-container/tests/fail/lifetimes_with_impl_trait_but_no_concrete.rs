use auto_impl::auto_impl;

#[auto_impl(&)]
trait Config {
    fn port(&self) -> u16;
}

struct EnvConfig;

impl Config for EnvConfig {
    fn port(&self) -> u16 {
        8080
    }
}

#[auto_impl(&)]
trait DAL {}

struct PostgresDAL;

impl DAL for PostgresDAL {}

struct Service<C, D> {
    config: C,
    dal: D,
}

impl<C: Config, D: DAL> Service<C, D> {
    fn new(config: C, dal: D) -> Self {
        println!("Service started on port {}", config.port());
        Self { config, dal }
    }
}

#[despatma_dependency_container::dependency_container]
impl DependencyContainer {
    #[Scoped]
    fn config(&self) -> impl Config {
        EnvConfig
    }

    #[Singleton]
    fn dal(&self) -> impl DAL {
        PostgresDAL
    }

    fn service(&self, config: impl Config, dal: impl DAL) -> Service<impl Config, impl DAL> {
        Service::new(config, dal)
    }
}

fn main() {
    let container = DependencyContainer::new();
    let _service = container.service();
}
