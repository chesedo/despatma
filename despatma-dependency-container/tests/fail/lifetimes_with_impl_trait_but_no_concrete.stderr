error: Need to know which type to store to manage this scoped dependency

         = help: Consider adding a type hint to the lifetime attribute
         = note: #[Scoped(ScopedType)]

  --> tests/fail/lifetimes_with_impl_trait_but_no_concrete.rs:38:25
   |
38 |     fn config(&self) -> impl Config {
   |                         ^^^^^^^^^^^

error: Need to know which type to store to manage this singleton dependency

         = help: Consider adding a type hint to the lifetime attribute
         = note: #[Singleton(SingletonType)]

  --> tests/fail/lifetimes_with_impl_trait_but_no_concrete.rs:43:22
   |
43 |     fn dal(&self) -> impl DAL {
   |                      ^^^^^^^^

error[E0433]: failed to resolve: use of undeclared type `DependencyContainer`
  --> tests/fail/lifetimes_with_impl_trait_but_no_concrete.rs:53:21
   |
53 |     let container = DependencyContainer::new();
   |                     ^^^^^^^^^^^^^^^^^^^ use of undeclared type `DependencyContainer`
